### PR TITLE
Fail fast if managed task workers are near-OOM

### LINKED ISSUE #1633: Excessive fuzzer memory usage can cause confusing task failures
Fuzzers (and user `target_exe`'s in general) can hang or OOM. We detect hangs with timeouts, but OOMs can take down a node and/or fail a task due to cross-process OOM-ing.

This can be true even in LibFuzzers, where the `-rss_limit_mb` arg results in a [periodic check](https://github.com/llvm/llvm-project/blob/23091f7d504afde4bba3fc885718a1633746e063/compiler-rt/lib/fuzzer/FuzzerDriver.cpp#L304-L311) of memory usage, [triggering a callback](https://github.com/llvm/llvm-project/blob/23091f7d504afde4bba3fc885718a1633746e063/compiler-rt/lib/fuzzer/FuzzerLoop.cpp#L308-L321) on excess usage.

The core problem here is that we need to be able to quickly differentiate OneFuzz or other system bugs from faults in user code. Ideally, any input-triggered OOM should be captured as a faulting input. Failing that, user-provided targets should not be able to indirectly fail a node. In the worst case, the task should still fail cleanly and informatively.

Possible solutions:
- Track _total_ system memory usage, and informatively fast-fail the task if it exceeds some limit
- Run all user-provided binaries with resource limits, via rlimit/job objects

### PR DESCRIPTION
**Pull request by ranweiler**
### Summary
- Add `onefuzz::memory::available_bytes()` to enable checking system-wide memory usage
- In managed task worker runs, heuristically check for imminent OOM conditions and try to exit early

### Testing
- [x] Added unit tests for limited parsing of `/proc/meminfo` on Linux
- [x] Memory querying functionally tested locally on both OSes using `memory` example binary in `onefuzz` crate
- [x] Near-OOM check in `onefuzz-agent` functionally tested by forcing external on-VM OOM
  - [x] Linux
  - [x] Windows

Tested on Windows by using `notmyfault` from SysInternals to leak from the paged memory pool, on a node where a healthy task was in the `running` state. On Linux, disabled OOM-killer, set overcommit to always-allow, and ran small custom program that rapidly leaks heap memory. Checked for task failure, with the reason matching the new error message, as well as telemetry.

Closes #1633.

### CODE REVIEW

File: src/agent/onefuzz-agent/src/managed/cmd.rs [hunk 1]
```diff
@@ -23,6 +39,35 @@ pub async fn run(args: &clap::ArgMatches<'_>) -> Result<()> {
     result
 }
 
+// Periodically check available system memory.
+//
+// If available memory drops below the minimum, exit informatively.
+async fn out_of_memory(min_bytes: u64) -> Result<OutOfMemory> {
```
Comment (ranweiler) [inline src/agent/onefuzz-agent/src/managed/cmd.rs:45]: I could have use the private named constant that we always pass here. I only parameterized this check to make it clear how we'd make  it configurable (and thread the exact limit through to the final error message).

File: src/agent/onefuzz/examples/memory.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+fn main() {
```
Comment (ranweiler) [inline src/agent/onefuzz/examples/memory.rs:1]: This is just an example binary for manual testing.

File: src/agent/onefuzz-agent/src/managed/cmd.rs (hunk 1 above)
Comment (stishkin) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:45]: add that as comment in the code ?

File: src/agent/onefuzz-agent/src/managed/cmd.rs [hunk 3]
```diff
@@ -13,7 +19,17 @@ pub async fn run(args: &clap::ArgMatches<'_>) -> Result<()> {
     let config = Config::from_file(config_path, setup_dir)?;
 
     init_telemetry(config.common());
-    let result = config.run().await;
+
+    let result = tokio::select! {
+        result = config.run() => result,
+
+        // Ignore this task if it returns due to a querying error.
+        Ok(oom) = out_of_memory(MIN_AVAILABLE_BYTES) => {
```
Comment (ranweiler) [inline src/agent/onefuzz-agent/src/managed/cmd.rs:27]: @chkeita, note how this differs a bit from our earlier discussion.

The supervisor `Agent` is implemented using (dynamic trait object) indirection to allow integration testing with test doubles for its internal components. Trying to add this global resource check as-is would pollute that, and was actually a little awkward (the supervisor really needs to be refactored to be `select`-driven).

This was the most minimal alternative way I could think up to get the behavior change I wanted, which was: "if we are in the `busy` state, running a task, and we're about to OOM, stop the task with an informative error".

File: src/agent/onefuzz-agent/src/managed/cmd.rs (hunk 1 above)
Comment (ranweiler) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:45]: IMO the actual usage is clear enough, I was more just noting it here for discussion's sake.

Do we think we'll want to make it tunable (i.e. per-node or task) in the future? If not, I can just simplify it by making it NOT parameterized.

Comment (ranweiler) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:48]: (just added a comment anyway)

Comment (mgreisen) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:52]: I'm ok with leaving it tunable.

**Review by mgreisen — APPROVED** [review]

File: src/agent/onefuzz-agent/src/managed/cmd.rs [hunk 4]
```diff
@@ -23,6 +43,43 @@ pub async fn run(args: &clap::ArgMatches<'_>) -> Result<()> {
...
+                        available_bytes,
+                        min_bytes,
+                    });
+                }
+            }
+            Err(err) => {
+                warn!("error querying system memory usage: {}", err);
+                return Err(err);
+            }
+        }
```
Comment (chkeita) [inline src/agent/onefuzz-agent/src/managed/cmd.rs:67]: This seems a bit agressive to stop everything as soon as we are unable to read the memory.
I would suggest raising the error if we fail to read the memory after a small number of attempts

File: src/agent/onefuzz-agent/src/managed/cmd.rs [hunk 5]
```diff
@@ -13,7 +23,17 @@ pub async fn run(args: &clap::ArgMatches<'_>) -> Result<()> {
...
-    let result = config.run().await;
+
+    let result = tokio::select! {
+        result = config.run() => result,
+
+        // Ignore this task if it returns due to a querying error.
+        Ok(oom) = out_of_memory(MIN_AVAILABLE_BYTES) => {
+            // Convert the OOM notification to an error, so we can log it below.
+            let err = format_err!("out of memory: {} bytes available, {} required", oom.available_bytes, oom.min_bytes);
+            Err(err)
```
Comment (chkeita) [inline src/agent/onefuzz-agent/src/managed/cmd.rs:34]: I think the decision to stop the task when OOM should be configurable and enabled by default.

File: src/agent/onefuzz-agent/src/managed/cmd.rs (hunk 4 above)
Comment (ranweiler) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:67]: For the record, the current behavior only stops the memory watchdog async task. But agreed, we should only give up the memory check after several failures in a row.

File: src/agent/onefuzz-agent/src/managed/cmd.rs (hunk 5 above)
Comment (ranweiler) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:36]: Added to the common config, on by default: 46a3d7a899ac90049fab00a3f04d1f3e71e4c222

Can be disabled by [setting to 0](https://github.com/microsoft/onefuzz/pull/1657/commits/46a3d7a899ac90049fab00a3f04d1f3e71e4c222#diff-083902b0f7ce9dceb9ecfb4eb6dda1de4a3595bd2161c306505e925880271b1cR55-R57).

Comment (ranweiler) [reply src/agent/onefuzz-agent/src/managed/cmd.rs:36]: Added: https://github.com/microsoft/onefuzz/pull/1657/commits/0befdc5f5608141a0e8fc86122dc0b37e2c432b0

**Review by chkeita — APPROVED** [review]

**Comment by ranweiler** [comment]
- Add `onefuzz::memory::available_bytes()` to enable checking system-wide memory usage
- In managed task worker runs, heuristically check for imminent OOM conditions and try to exit early

**Pull request merged**